packages/ui/src/components/Visualization/ContextToolbar/ExportDocument/ExportDocumentPreviewModal.tsx

```diff
@@ -0,0 +1,163 @@
+import {
+  Button,
+  Form,
+  FormGroup,
+  ModalBody,
+  ModalHeader,
+  ModalVariant,
+  Split,
+  SplitItem,
+  TextInput,
+  Toolbar,
+  ToolbarItem,
+} from '@patternfly/react-core';
+import { FunctionComponent, useContext, useEffect, useState } from 'react';
+import { EntitiesMenu } from './EntitiesMenu';
+import { Modal } from '@patternfly/react-core';
+import Markdown from 'react-markdown';
+import { DocumentationEntity } from '../../../../models/documentation';
+import { EntitiesContext, VisibleFlowsContext } from '../../../../providers';
+import { DocumentationService } from '../../../../services/documentation.service';
+import { DownloadIcon } from '@patternfly/react-icons';
+import { markdownComponentMapping } from './MarkdownComponentMapping';
+import remarkGfm from 'remark-gfm';
+import { Element } from 'hast';
+
+type IExportDocumentPreviewModal = {
+  isOpen?: boolean;
+  onClose: () => void;
+};
+
+export const ExportDocumentPreviewModal: FunctionComponent<IExportDocumentPreviewModal> = ({
+  isOpen = true,
+  onClose,
+}) => {
+  const fileNameBase = 'route-export';
+  const { entities, visualEntities } = useContext(EntitiesContext)!;
+  const { visibleFlows, visualFlowsApi } = useContext(VisibleFlowsContext)!;
+  const [markdownText, setMarkdownText] = useState<string>('');
+  const [flowImageBlob, setFlowImageBlob] = useState<Blob>();
+  const [downloadFileName, setDownloadFileName] = useState<string>(fileNameBase + '.zip');
+  const [isModalOpen, setIsModalOpen] = useState<boolean>(isOpen);
+  const initialDocEntities = DocumentationService.getDocumentationEntities(entities, visualEntities, visibleFlows);
+  const [documentationEntities, setDocumentationEntities] = useState<DocumentationEntity[]>(initialDocEntities);
+
+  const onUpdateDocumentationEntities = (documentationEntities: DocumentationEntity[]) => {
+    documentationEntities.forEach((docEntity) => {
+      if (
+        docEntity.isVisualEntity &&
+        ((docEntity.isVisible && !visibleFlows[docEntity.entity!.id]) ||
+          (!docEntity.isVisible && visibleFlows[docEntity.entity!.id]))
+      ) {
+        visualFlowsApi.toggleFlowVisible(docEntity.entity!.id);
+      }
+    });
+
+    setDocumentationEntities([...documentationEntities]);
+  };
+
+  useEffect(() => {
+    const updatePreview = async () => {
+      let imageUrl = '';
+      DocumentationService.generateFlowImage()
+        .then((imageBlob) => {
+          if (imageBlob) {
+            setFlowImageBlob(imageBlob);
+            imageUrl = window.URL.createObjectURL(imageBlob);
+          }
+        })
+        .catch((error) => {
+          console.error(error);
+        })
+        .finally(() => {
+          const md = DocumentationService.generateMarkdown(documentationEntities, imageUrl);
+          setMarkdownText(md);
+        });
+    };
```

Comment:
We can spare the `async` calificator because we're not using `async/await` syntax, also, it would be nice to return a cleanup function, so if the user navigates somewhere else, this function gets cancelled.

```suggestion
  useEffect(() => {
    const updatePreview = () => {
      let imageUrl = '';
      DocumentationService.generateFlowImage()
        .then((imageBlob) => {
          if (imageBlob) {
            setFlowImageBlob(imageBlob);
            imageUrl = window.URL.createObjectURL(imageBlob);
          }
        })
        .catch((error) => {
          console.error(error);
        })
        .finally(() => {
          const md = DocumentationService.generateMarkdown(documentationEntities, imageUrl);
          setMarkdownText(md);
        });
    };

    // A workaround for React 18 synchronous useEffect execution, which causes the flow image
    // to be captured before the entity visibility reflects
    // https://react.dev/reference/react/useEffect#caveats
    const clearTimer = setTimeout(updatePreview, 0);
    return () => {
      clearTimeout(clearTimer);
    };
  }, [documentationEntities, visibleFlows]);
```